web/api/v1/api.go

```diff
@@ -849,50 +849,80 @@ type recordingRule struct {
 func (api *API) rules(r *http.Request) apiFuncResult {
 	ruleGroups := api.rulesRetriever.RuleGroups()
 	res := &RuleDiscovery{RuleGroups: make([]*RuleGroup, len(ruleGroups))}
+	typeParam := strings.ToLower(r.URL.Query().Get("type"))
+
+	if typeParam != "" && typeParam != "alert" && typeParam != "record" {
+		level.Warn(api.logger).Log("msg", "Invalid type param recieved. Will send all rules.")
+	}
+
 	for i, grp := range ruleGroups {
 		apiRuleGroup := &RuleGroup{
 			Name:     grp.Name(),
 			File:     grp.File(),
 			Interval: grp.Interval().Seconds(),
 			Rules:    []rule{},
 		}
-
 		for _, r := range grp.Rules() {
 			var enrichedRule rule
 
 			lastError := ""
 			if r.LastError() != nil {
 				lastError = r.LastError().Error()
 			}
-
-			switch rule := r.(type) {
-			case *rules.AlertingRule:
-				enrichedRule = alertingRule{
-					Name:        rule.Name(),
-					Query:       rule.Query().String(),
-					Duration:    rule.Duration().Seconds(),
-					Labels:      rule.Labels(),
-					Annotations: rule.Annotations(),
-					Alerts:      rulesAlertsToAPIAlerts(rule.ActiveAlerts()),
-					Health:      rule.Health(),
-					LastError:   lastError,
-					Type:        "alerting",
+			if typeParam == "alert" {
+				alert, ok := r.(*rules.AlertingRule)
+				if ok {
+					apiRuleGroup.Rules = append(apiRuleGroup.Rules, alertingRule{
```

Comment:
Hm, this is a lot of code duplication, and we should probably still keep the `Type` field even when we are filtering. Can we just keep the existing switch statement and add `break` conditions there? Like:

```golang
switch rule := r.(type) {
case *rules.AlertingRule:
	if !returnAlerts {
		break
	}
	// ...
case *rules.RecordingRule:
	if !returnRecording {
		break
	}
	// ...
}
```

...and define the `returnRecording` and `returnAlerts` vars above (similar to https://github.com/prometheus/prometheus/blob/46bbac20126ed43654eb256368b618eaa86b3aa3/web/api/v1/api.go#L612-L613), so we don't have to rely on string values here anymore.

Reply:
Yeah, good idea, thanks!